rclcpp/include/rclcpp/client.hpp

```diff
@@ -195,6 +201,88 @@ class Client : public ClientBase
 
   RCLCPP_SMART_PTR_DEFINITIONS(Client)
 
+  /// A convenient Client::Future and request id pair.
+  class FutureAndRequestId
+  {
+public:
+    FutureAndRequestId(Future impl, int64_t req_id)
+    : impl_(std::move(impl)), req_id_(req_id)
+    {}
+
+    /// Allow implicit conversions to `std::future` by reference.
+    // TODO(ivanpauno): Maybe, deprecate this in favor of get_future() (?)
+    operator Future &() {return impl_;}
+
+    /// Deprecated, use take_future() instead.
+    /**
+     * Allow implicit conversions to `std::future` by value.
+     * \deprecated
+     */
+    [[deprecated("FutureAndRequestId: use take_future() instead of an implicit conversion")]]
+    operator Future() {return impl_;}
+
+    /// Returns the internal `std::future`, moving it out.
+    /**
+     * After calling this method, the internal future gets invalidated.
+     */
+    Future
+    take_future() {return impl_;}
+
+    /// Getter for the internal future.
+    Future &
+    get_future() {return impl_;}
+
+    /// Getter for the internal future.
+    const Future &
+    get_future() const {return impl_;}
+
+    /// Returns the request id associated with this future.
+    int64_t get_request_id() const {return req_id_;}
+
+    // delegate future like methods in the std::future impl_
+
+    /// See std::future::share().
+    SharedFuture share() noexcept {return impl_.share();}
+    /// See std::future::get().
+    SharedResponse get() {return impl_.get();}
+    /// See std::future::valid().
+    bool valid() const noexcept {return impl_.valid();}
+    /// See std::future::wait().
+    void wait() const {return impl_.wait();}
+    /// See std::future::wait_for().
+    template<class Rep, class Period>
+    std::future_status wait_for(
+      const std::chrono::duration<Rep, Period> & timeout_duration) const
+    {
+      return impl_.wait_for(timeout_duration);
+    }
+    /// See std::future::wait_until().
+    template<class Clock, class Duration>
+    std::future_status wait_until(
+      const std::chrono::time_point<Clock, Duration> & timeout_time) const
+    {
+      return impl_.wait_until(timeout_time);
+    }
+
+    // Rule of five, we could use the rule of zero here, but better be explicit as some of the
+    // methods are deleted.
+
+    /// Move constructor.
+    FutureAndRequestId(FutureAndRequestId && other) noexcept = default;
+    /// Deleted copy constructor, each instance is a unique owner of the future.
+    FutureAndRequestId(const FutureAndRequestId & other) = delete;
+    /// Move assignment.
+    FutureAndRequestId & operator=(FutureAndRequestId && other) noexcept = default;
+    /// Deleted copy assignment, each instance is a unique owner of the future.
+    FutureAndRequestId & operator=(const FutureAndRequestId & other) = delete;
+    /// Destructor.
+    ~FutureAndRequestId() = default;
+
+private:
+    Future impl_;
+    int64_t req_id_;
```

Comment:
I think we can make this two members public, this is simply a pair there's no invariant here.

The future like methods were added to avoid breaking pre-existing code.
It's not a bad idea to deprecate all of them though, that way we don't break pre-existing code but users get a warning and they take a look to `remove_pending_request()`.